certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,372 @@
+            requireInvariant countZeroIffListEmpty();
+        }
+    }
+
+// reach encodes a linear order. This axiom corresponds to Table 2 in [1].
+invariant reachInvariant()
+    forall address wallet. forall address X. forall address Y. forall address Z. (
+        reach(wallet, X, X)
+        && (reach(wallet, X, Y) && reach(wallet, Y, X) => X == Y)
+        && (reach(wallet, X, Y) && reach(wallet, Y, Z) => reach(wallet, X, Z))
+        && (reach(wallet, X, Y) && reach(wallet, X, Z) => (reach(wallet, Y, Z) || reach(wallet, Z, Y)))
+    )
+    {
+        preserved {
+            requireInvariant reachNull();
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+        }
+    }
+
+// every element with non-zero guardian field is reachable from SENTINEL (head of the list)
+invariant inListReachable()
+    (forall address wallet. forall address key. ghostGuardians[wallet][key] != 0 => reach(wallet, SENTINEL, key))
+    {
+        preserved {
+            requireInvariant reachInvariant();
+            requireInvariant reachNull();
+            requireInvariant reachableInList();
+            requireInvariant countZeroIffListEmpty();
+            requireInvariant emptyListNotReachable();
+        }
+    }
+
+invariant reachHeadNext()
+    forall address wallet. forall address X. (reach(wallet, SENTINEL, X) && X != SENTINEL && X != NULL) =>
+           (ghostGuardians[wallet][SENTINEL] != SENTINEL && reach(wallet, ghostGuardians[wallet][SENTINEL], X))
+    {   
+        preserved {
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+            requireInvariant reachInvariant();
+            requireInvariant reachNull();
+            requireInvariant countZeroIffListEmpty();
+            requireInvariant emptyListNotReachable();
+            requireInvariant nextNull();
+        }
+    }
+
+// every element reaches its direct successor (except for the tail-SENTINEL).
+invariant reachNext()
+    forall address wallet. forall address X. reachSucc(wallet, X, ghostGuardians[wallet][X])
+    {
+        preserved {
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+            requireInvariant reachNull();
+            requireInvariant reachInvariant();
+        }
+    }
+
+// Express the next relation from the reach relation by stating that it is reachable and there is no other element
+// in between.
+// This is equivalent to P_next from Table 3.
+definition isSucc(address wallet, address a, address b) returns bool = reach(wallet, a, b) && a != b && (forall address Z. reach(wallet, a, Z) && reach(wallet, Z, b) => (a == Z || b == Z));
+definition nextOrNull(address n) returns address = n == SENTINEL ? NULL : n;
+
+// Invariant stating that the guardians storage pointers correspond to the next relation, except for the SENTINEL tail marker.
```

Comment:
This is not an `invariant` right?

Reply:
https://github.com/5afe/CandideWalletContracts/pull/20/commits/9f099129534070db847f1123dad3c51e04931c9f